File: matlab/part_b_example.m
% =========================================================================
% Example usage of System1 class in MATLAB
% Make sure System1.p and is in the current folder or path.
% =========================================================================
clc;clear;
%% --------------------- SYSTEM 1 USAGE ----------------------------------

s2 = System2(42); % Last two digits of student ID (last digit if it is 0x)

% Step response
[~, t2, y2] = s2.step();
figure;
plot(t2, y2);
title('System2 - Step Response');
grid on;

% Ramp response
[~, t2, y2] = s2.ramp();
figure;
plot(t2, y2);
title('System2 - Ramp Response');
grid on;

% Arbitrary input sequence
s2 = s2.reset(); % Reset internal state
input_seq = linspace(0, 1, 100);
output_seq = zeros(size(input_seq));

for i = 1:length(input_seq)
    [s2, output_seq(i)] = s2.output(input_seq(i));
end

figure;
plot(input_seq, output_seq);
title('System2 - Arbitrary Input Test');
grid on;

File: python/part_b_example.py
import sys, pathlib, importlib.util
import numpy as np
import matplotlib.pyplot as plt

# Example usage of the System2 class. If you write your own code, copy the 
# part below.
# You need Python 3.8 to use the compiled module.

############################ Load the system class ############################
pyc_path = pathlib.Path("./") / "python/System2.cpython-38.pyc"
spec = importlib.util.spec_from_file_location("system2", pyc_path)
module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(module)

System2 = module.System2
##############################################################################

s = System2(35) # Last two digits of your student ID (last digit if it is 0x)

# 1 - Step input
s.reset()
t, y = s.step() # Optionally, you can give t_final input

plt.figure()
plt.plot(t, y)
plt.title("Step Response")
plt.grid()
plt.show()

# 2 - Ramp input results
s.reset()
t, y = s.ramp() # Optionally, you can give t_final input

plt.figure()
plt.plot(t, y)
plt.title("Ramp Response ")
plt.grid()
plt.show()

# 3 - Test with arbitrary input sequence
# This system takes one input at a time, so we'll simulate a sequence
s.reset() # Dont forget to reset
inputs = np.linspace(0, 1, 100)
outputs = [s.output(float(u)) for u in inputs]

plt.figure()
plt.plot(inputs, outputs)
plt.title("System Output for Varying Input")
plt.grid()
plt.show()
